File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int lower_bound(vector<int> &arr , int n , int target){
        int low = 0 ;
        int high = n -1 ; 
        int ans = n ;
        while(low <= high){
            int mid = low+(high - low)/2; 
            if(arr[mid] >= target){
                ans = mid ; 
                high = mid - 1; 
            }
            else{
                low = mid + 1 ; 
            }
        } 
        return ans ; 
    }
    int upper_bound(vector<int> &arr , int n , int target){
        int low = 0 ;
        int high = n - 1 ; 
        int ans = n ;
        while(low <= high){
            int mid = low+(high - low)/2; 
            if(arr[mid] == target){
                ans = mid ; 
                low = mid + 1 ;  
            }
            else if(arr[mid]< target){
                low = mid + 1; 
            }
            else{
                high = mid - 1 ; 
            }
        } 
        return ans ; 
    }
    vector<int> searchRange(vector<int>& arr, int target) {
        int n = arr.size() ; 
        int lower = lower_bound(arr , n  , target) ;
        int upper = upper_bound(arr , n , target) ; 
        if(lower == n || arr[lower] != target){
            return {-1 , -1 }; 
        }
        return{lower , upper}; 
    }
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public static int[] findBounds(int[] arr, int n, int target) {
        int low = 0;
        int high = n - 1;
        int lower = n, upper = n;

        // lower bound
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] >= target) {
                lower = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        low = 0; 
        high = n - 1;

        //upper bound 
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == target) {
                upper = mid;
                low = mid + 1;
            } else if (arr[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return new int[]{lower, upper};
    }

    public int[] searchRange(int[] arr, int target) {
        int n = arr.length;
        int[] bounds = findBounds(arr, n, target);
        int lower = bounds[0];
        int upper = bounds[1];
        
        if (lower == n || arr[lower] != target) {
            return new int[]{-1, -1};
        }
        return new int[]{lower, upper};
    }
}
